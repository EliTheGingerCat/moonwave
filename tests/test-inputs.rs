use std::path::Path;
use std::process::{Command, Stdio};

#[test]
fn everything_sandwich() -> anyhow::Result<()> {
    run_moonwave("passing/everything_sandwich.lua", 0)
}

#[test]
fn class_with_function() -> anyhow::Result<()> {
    run_moonwave("passing/class_with_function.lua", 0)
}

#[test]
fn fabric() -> anyhow::Result<()> {
    run_moonwave("passing/fabric.lua", 0)
}

#[test]
fn failing_function() -> anyhow::Result<()> {
    run_moonwave("failing/function.lua", 1)
}

#[test]
fn failing_function_no_within() -> anyhow::Result<()> {
    run_moonwave("failing/function_no_within.lua", 1)
}

fn run_moonwave(file_name: &str, expected_status: i32) -> anyhow::Result<()> {
    let path = Path::new("test-input").join(file_name);

    let child = Command::new(env!("CARGO_BIN_EXE_moonwave"))
        .arg("extract")
        .arg(path)
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .env("NO_COLOR", "1")
        .spawn()?;

    let output = child.wait_with_output()?;
    let stdout = String::from_utf8(output.stdout)?;
    let stderr = String::from_utf8(output.stderr)?;

    let stdout_name = format!("{}-stdout", file_name);
    let stderr_name = format!("{}-stderr", file_name);

    insta::assert_snapshot!(stdout_name, stdout);
    insta::assert_snapshot!(stderr_name, stderr);

    assert_eq!(output.status.code(), Some(expected_status));

    Ok(())
}
