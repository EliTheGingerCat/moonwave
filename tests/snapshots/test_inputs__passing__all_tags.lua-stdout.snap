---
source: tests/test-inputs.rs
expression: stdout

---
[
  {
    "functions": [
      {
        "name": "comingSoon",
        "desc": "",
        "params": [],
        "returns": [],
        "function_type": "static",
        "unreleased": true,
        "source": {
          "line": 50,
          "path": ""
        }
      }
    ],
    "properties": [
      {
        "name": "notReady",
        "desc": "",
        "lua_type": "boolean",
        "unreleased": true
      }
    ],
    "types": [],
    "name": "Bar",
    "desc": "",
    "unreleased": true
  },
  {
    "functions": [
      {
        "name": "new",
        "desc": "This function creates a new Foo",
        "params": [
          {
            "name": "a",
            "desc": "param a",
            "lua_type": "A A A"
          },
          {
            "name": "b",
            "desc": "param b",
            "lua_type": "B B B"
          }
        ],
        "returns": [
          {
            "desc": "return a",
            "lua_type": "a"
          },
          {
            "desc": "return b",
            "lua_type": "b"
          }
        ],
        "function_type": "static",
        "tags": [
          "uno",
          "dos"
        ],
        "errors": [
          {
            "lua_type": "c",
            "desc": "this errors sometimes. shrug"
          }
        ],
        "realm": [
          "Client",
          "Server"
        ],
        "since": "v1.23",
        "deprecated": {
          "version": "v1.5",
          "desc": "Use something else"
        },
        "private": true,
        "yields": true,
        "ignore": true,
        "source": {
          "line": 45,
          "path": ""
        }
      }
    ],
    "properties": [
      {
        "name": "ready",
        "desc": "",
        "lua_type": "boolean",
        "tags": [
          "salad",
          "fries",
          "nuggets"
        ],
        "realm": [
          "Client",
          "Server"
        ],
        "since": "v0.1.1",
        "deprecated": {
          "version": "v1.3",
          "desc": "Use blah"
        },
        "private": true,
        "readonly": true,
        "ignore": true
      }
    ],
    "types": [
      {
        "name": "nilOrNumber",
        "desc": "Nil or number matey",
        "lua_type": "nil | number",
        "tags": [
          "arrgh",
          "yarr"
        ],
        "private": true,
        "ignore": true
      },
      {
        "name": "Command",
        "desc": "An object describing a command",
        "fields": [
          {
            "name": "Name",
            "lua_type": "string",
            "desc": "the name of the command"
          },
          {
            "name": "Groups",
            "lua_type": "array<string>",
            "desc": "A list of groups that the command contains"
          },
          {
            "name": "Recursion",
            "lua_type": "Command",
            "desc": "This breaks the universe"
          }
        ],
        "tags": [
          "cmdr",
          "lua"
        ],
        "private": true,
        "ignore": true
      }
    ],
    "name": "Foo",
    "desc": "Here's a description for you\n\n::: info\nwith an admonition\n:::",
    "tags": [
      "bark",
      "meow",
      "rawr"
    ],
    "realm": [
      "Client",
      "Server",
      "Plugin"
    ],
    "deprecated": {
      "version": "v1.2",
      "desc": "Use x instead"
    },
    "since": "v1.0",
    "private": true,
    "ignore": true
  }
]

